Adding Python3 to OSX

CursedRock17 opened the pull request:
<!-- Thank you for contributing a change to the rosdistro. There are two primary types of submissions.
Please select the appropriate template from below: ROSDEP_RULE_TEMPLATE or DOC_INDEX_TEMPLATE

If you're making a new release with bloom please use bloom to create the pull request automatically.
If you've already run the release bloom has a `--pull-request-only` option you can use.-->

<!-- ROSDEP_RULE_TEMPLATE: Submitter Please review the contributing guidelines: https://github.com/ros/rosdistro/blob/master/CONTRIBUTING.md -->

Adding the Python3 package to the OSX version of rosdistro, as it is already required with the ubuntu version and there is a homebrew package.

## Package name:

python3

## Package Upstream Source:

https://www.python.org/

## Purpose of using this:

There are libraries like moveit_ros_planning_interface that utilize python3 as part of rosdep, but it currently doesn't have access. Although the package has a different name it is already on homebrew which makes this simple to add.


<!-- Replace the REQUIRED areas with the URL to the package.  For IF AVAILABLE areas, either put in the URL to the package or state 'not available'.
More info at https://github.com/ros/rosdistro/blob/master/CONTRIBUTING.md#guidelines-for-rosdep-rules -->

- macOS: https://formulae.brew.sh/formula/python@3.11#default

CursedRock17 committed:
Added libomp-dev for osx

CursedRock17 committed:
Adding Python3 to OSX

tfoote commented:
Please rebase this. You appear to have added it on top of #36746

CursedRock17 committed:
Revert "Added libomp-dev for osx"

This reverts commit d98f3b3e9723babb3cd9c377242efd9b5a48c282.

nuclearsandwich commented on rosdep/osx-homebrew.yaml:
```diff
@@ -691,6 +691,10 @@ python-vtk:
+python3:
+  osx:
+    homebrew:
+      packages: []
```
Is the package listed intended to be empty?

CursedRock17 replied:
No, but I went through and recertified different versions to make it load correctly on python3.

nuclearsandwich replied:
Looking at https://docs.brew.sh/Homebrew-and-Python#python-3y and https://formulae.brew.sh/formula/python@3.11

It seems like the `python@3` or `python3` alias for the latest python3.x formula is an appropriate definition here.

```suggestion
      packages: [python3]
```

CursedRock17 committed:
Adding certain python version osx-homebrew.yaml

nuclearsandwich commented on rosdep/osx-homebrew.yaml:
```diff
@@ -650,7 +650,7 @@ protobuf-dev:
   osx:
     homebrew:
-      packages: []
+      packages: [python, python3]
```
Looking at https://docs.brew.sh/Homebrew-and-Python#python-3y and https://formulae.brew.sh/formula/python@3.11

I actually think we should make the `python` rosdep key, which is implicitly python2, undefined on homebrew, since python2 has been removed.

CursedRock17 replied:
Is there an argument to remove it then, or will it just provide users a warning?

nuclearsandwich replied:
The change as proposed is incorrect as both the `python` and `python3` formula names will result in python3 being installed. By removing the definition, any package relying on `python` (implicitly python2, and no longer available on macOS) will see a failure to resolve the rosdep key. Which is the preferred behavior if a dependency is not resolvable.

nuclearsandwich replied:
I pushed [3ae1423](https://github.com/ros/rosdistro/pull/36747/commits/3ae14238aa26b2c3796b169f90def664593a5088) implementing my intended suggestion. There is a difference in rosdep between an "undefined" definition and an "empty" definition. An empty package list means that installing no packages satisfies this key. It is used in situations where the packages listed are part of the core system and therefore not installed by any specific package and other special situations where a package manager may actually be responsible for managing a package. To make the error condition we want for this key on macOS, we have to either assign it a value of `null` or, as I've done here, remove the definition entirely since there are no other definitions for this key in this file.

CursedRock17 committed:
Making python homebrew package undefined

nuclearsandwich committed:
Remove python key from osx entirely.

This key implicitly resolves python2, which is no longer available on
macOS via homebrew or otherwise.

Removing it will produce failures to resolve this key on that platform,
which is an intended and desired result.

The empty definition which is removed here essentially means that this
dependency is resolved by installing no packages rather than this
dependency has no definition. This would work since macOS maintained
a system installation of python2.

CursedRock17 committed:
Adding dlib
